### PR TITLE
Migrate PoW to PoS history proactively

### LINKED ISSUE #2355: Migration: Migrate existing history proactively
Do not wait until the activation window to start migrating the history: the majority of it is not going to change so we can eagerly migrate the history if the tool is started before the activation window.

### PR DESCRIPTION
**Pull request by Eligioo**
## What's in this pull request?
This PR adds the functionality for the `nimiq-pow-migration` client to start the PoW to PoS migration before the pre-stake window is closed. Additionally it is responsible for continuing to migrate the PoW history if instructed when for example a new activation window is started.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: pow-migration/src/main.rs [hunk 1]
```diff
@@ -233,10 +245,37 @@ async fn main() {
...
+            }
+            sleep(Duration::from_secs(60)).await;
+        }
+
         let genesis_config;
 
-        // Do the migration
         loop {
+            let pos_history_store_height = rx_migration_completed.borrow();
+            // Wait for the Pow to PoS migration to be caught up with the canidate block
```
Comment (hrxi) [inline pow-migration/src/main.rs]:
```suggestion
            // Wait for the PoW to PoS migration to be caught up with the candidate block
```

Comment (Eligioo) [reply pow-migration/src/main.rs]: @hrxi fixed

File: pow-migration/src/history/mod.rs [hunk 2]
```diff
@@ -102,100 +105,128 @@ fn from_pow_transaction(pow_transaction: &PoWTransaction) -> Result<Transaction,
...
+                        break;
+                    }
 
-                    assert_eq!(
-                        pow_transaction.hash,
-                        pos_transaction.hash::<Blake2bHash>().to_hex()
+                    pow_head_height = pow_client.block_number().await.unwrap();
+                    log::info!(
+                        "Waiting for block {} to be confirmed before we can migrate it",
+                        block_height
```
Comment (jsdanielh) [inline pow-migration/src/history/mod.rs]:
```suggestion
                        block=block_height, "Waiting for block to be confirmed before we can migrate it",
                        
```

File: pow-migration/src/history/mod.rs [hunk 3]
```diff
@@ -102,100 +105,128 @@ fn from_pow_transaction(pow_transaction: &PoWTransaction) -> Result<Transaction,
...
+                .get_block_by_number(block_height, false)
+                .await
+                .unwrap();
+
+            // Get all transactions for this block height
+            let mut transactions = vec![];
+            let mut network_id = NetworkId::Main;
+            match block.transactions {
+                PoWTransactionSequence::BlockHashes(hashes) => {
+                    // Mark that we've migrated up until this block.
```
Comment (jsdanielh) [inline pow-migration/src/history/mod.rs]: I'd do this after we actually have done the migration for all of the hashes of this block height

File: pow-migration/src/main.rs [hunk 4]
```diff
@@ -205,21 +211,27 @@ async fn main() {
             }
         };
 
+        // Create channels in order to communicate with the PoW-to-PoS migrator
```
Comment (jsdanielh) [inline pow-migration/src/main.rs]:
```suggestion
        // Create channels in order to communicate with the PoW-to-PoS history migrator
```

File: pow-migration/src/main.rs [hunk 5]
```diff
@@ -233,10 +245,37 @@ async fn main() {
...
+            }
+            sleep(Duration::from_secs(60)).await;
+        }
+
         let genesis_config;
 
-        // Do the migration
         loop {
+            let pos_history_store_height = rx_migration_completed.borrow();
+            // Wait for the PoW to PoS migration to be caught up with the candidate block
```
Comment (jsdanielh) [inline pow-migration/src/main.rs]:
```suggestion
            // Wait for the PoW to PoS history migration to be caught up with the candidate block
```

File: pow-migration/src/main.rs [hunk 6]
```diff
@@ -233,10 +245,37 @@ async fn main() {
...
+            if *pos_history_store_height != candidate_block {
+                log::info!(
+                    "Waiting for the PoW history to be migrated up until candidate block {}, migrated up until block {}",
+                    candidate_block,
+                    *pos_history_store_height,
```
Comment (jsdanielh) [inline pow-migration/src/main.rs]:
```suggestion
                    candidate_block, current_migrated_block=*pos_history_store_height, "Waiting for the PoW history to be migrated up until candidate block",
```

File: pow-migration/src/main.rs [hunk 7]
```diff
@@ -255,6 +294,8 @@ async fn main() {
                     }
                     None => {
                         candidate_block += block_windows.readiness_window;
+                        // Instruct to migrate the PoW history up until the next candidate block
+                        tx_candidate_block.send(candidate_block).await.unwrap();
```
Comment (jsdanielh) [inline pow-migration/src/main.rs]: I think it might be better to just log the error and exit instead of unwraping here

**Comment by Eligioo** [comment]
Rebasing

**Comment by Eligioo** [comment]
@jsdanielh all changes have been resolved

**Review by jsdanielh — APPROVED** [review]
LGTM

**Comment by jsdanielh** [comment]
Rebasing it to merge it

**Pull request merged**